py/server/deephaven/table.py

```diff
@@ -694,6 +696,63 @@ def to_string(self, num_rows: int = 10, cols: Union[str, Sequence[str]] = None)
         except Exception as e:
             raise DHError(e, "table to_string failed") from e
 
+    def diff(self, table: Table, max_lines: int = 1, floating_inexact: bool = False,
+             floating_fraction: bool = False, ignore_column_order: bool = False) -> str:
+        """Returns the differences between this table and the provided table as a string.
+
+        This method starts by comparing the table sizes, and then the schema of the two tables, such as the number of
+        columns, column names, column types, column orders. If the schemas are different, the comparison stops and
+        the differences are returned. If the schemas are the same, the method proceeds to compare the data in the
+        tables. The method compares the data in the tables column by column and only records the first difference
+        found in each column.
+
+        Note, inexact comparison of floating numbers may sometimes be desirable due to their inherent imprecision.
+        When that is the case, the floating_inexact flag should be set to True and the absolute value of the
+        difference between two floating numbers will be compared against a threshold. The threshold is set to 0.0001
+        for Doubles and 0.005 for Floats. Only differences that are greater than the threshold are recorded. The
+        floating_fraction flag can be set to True to compare the fractional difference between two floating numbers
+        against the threshold. The fractional difference is calculated as the absolute difference divided by the
+        smaller one between the absolute values of the two numbers.
```

Comment:
As a numericist, the way the approximate equality is being done makes me very sad and is not very useful.
1. Julia's `isapprox` function is what we want to emulate.  It allows both absolute and relative tolerances to be checked at once.  It also allows NaN behavior to be set.  https://docs.julialang.org/en/v1/base/math/#Mathematical-Functions. https://www.jlhub.com/julia/manual/en/function/isapprox
2. The way our limits are set makes doing what I want impossible.

For limits:
1. We need absolute and relative limits.
2. We need a dict to supply by-column limits.
3. We need "default" limits that apply if no by-column limits are specified.

Reply:
This feels out of scope for this ticket. The issue I wrote was to expose our existing functionality, which is sufficiently helpful as to be the basis for most of the correctness testing in our unit test suite, and is fantastic for finding bugs in data or engine operations. If you would like to propose new requirements, please write a new ticket with those requirements. I do agree that the defaults seem arbitrary and simplistic. That said, I don't think that needs to be a high priority for us, nor do I think anyone is likely to use the per-column dictionary in the near term.

Reply:
I too realized that it was designed in a weirdly too limiting way. But in the spirit of treating it as an useful tool for debugging purpose (not production), I find a simple wrapping is tolerable. If we see the potential for this to become something much more than that, then I would propose to make it private first for internal use (including Core+).

Reply:
I created a ticket so that we can address this in the future.
https://github.com/deephaven/deephaven-core/issues/5852